### ISSUE TITLE
Eleven TrainerCallback modules import transformers at module scope (lazy factory called eagerly)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#320]
## The pattern

Eleven modules resolve their `TrainerCallback` base **at class-definition time**:

```python
def _get_trainer_callback_base():
    try:
        from transformers import TrainerCallback   # looks lazy
        return TrainerCallback
    except ImportError:
        return object

_TrainerCallbackBase = _get_trainer_callback_base()   # ...but CALLED at module scope

class RewardHackMitigationCallback(_TrainerCallbackBase):
    ...
```

The factory is lazy; the call is not. Importing any of these modules imports
`transformers` and therefore `torch` — measured at **~4.4 s** each:

| module | |
|---|---|
| `utils/reward_hack_control.py` | `_TrainerCallbackBase = ...` at module scope |
| `utils/reward_hacking.py` | same |
| `utils/echo_trap.py` | same |
| `utils/minillm.py` | same |
| `utils/rl_checkpoint.py` | same |
| `utils/relora.py` | `class ReLoRACallback(_try_import_callback_base())` |
| `utils/lisa.py` | same, inline in the class header |
| `monitoring/hf_push.py` | same |
| `monitoring/curriculum_callback.py` | same |
| `monitoring/grpo_stability_callback.py` | `_TrainerCallbackBase = ...` |
| `monitoring/callback.py` | leaks at import |

`trainer/raft.py:187` resolves the base **inside a factory function** and stays
clean (0.04 s) — that is the reference implementation already in the repo.

## Why it matters (and why it is not urgent)

On the trainer path this costs nothing: `transformers` is loaded anyway.

It becomes a user-visible regression the moment a **light** command imports one.
That is exactly what happened in v0.71.41: `commands/reward.py` imported
`utils/reward_stress`, which imported `utils/reward_hack_control` for a single
string constant, and `soup --help` went from **1.15 s to 6.0 s**.

That specific leak is fixed, and `tests/test_cli_startup_is_light.py` now asserts
the real invariant at runtime (`import soup_cli.cli` → `torch` not in
`sys.modules`), so a recurrence fails CI loudly instead of silently. This issue
is about the ten remaining latent modules.

## Why the existing guards did not catch it

Every one of these modules has a `test_no_top_level_torch`-style guard. They
parse the module source and assert it contains no top-level `import torch`.
That proves a **syntactic** property; the requirement is a **runtime** one. An
AST walk cannot see through a transitive import, nor through a lazily-written
factory that is called at module scope.

`tests/test_v07126.py::test_no_top_level_heavy_import_in_source` is the sharpest
example: it *imports* `reward_hack_control` (loading transformers, ~4.7 s) and
then asserts that the source has no top-level transformers import. The test
performs the very thing it certifies against, and passes.

## Proposed fix

Cheaper than it looks — **no `super()` call exists in any of the eleven class
bodies** (verified), so nothing needs re-indenting into a function. Two viable
shapes:

1. **PEP 562 module `__getattr__`** — build the class on first attribute access,
   cache it. `from mod import XCallback` and `isinstance` keep working.
2. **`type()` rebuild** — keep the body on a plain `object` base, synthesise the
   real subclass lazily.

Either way the acceptance bar is the same as #308: the resulting object must be a
real `transformers.TrainerCallback` subclass, because HF's `CallbackHandler`
dispatches every lifecycle event through `getattr(cb, event)(...)` with no
`hasattr` guard.

## Acceptance criteria

- [ ] All eleven modules import without pulling `transformers` / `torch`
- [ ] Each callback is still `isinstance(cb, TrainerCallback)` when transformers
      is installed (the #308 property — do not regress it)
- [ ] Each still inherits the no-op stubs for unimplemented events
      (`cb.on_epoch_begin(None, None, None)` must not raise)
- [ ] Extend `tests/test_cli_startup_is_light.py` with a runtime per-module check
      covering all eleven
- [ ] Retire or downgrade the per-file AST guards these modules carry — they
      certify a property they cannot observe
- [ ] Correct the `tests/test_issue308_callback_subclass.py` docstring claim that
      the factory "keeps the modules transformers-free at import time"

## Not in scope

Making these modules light has **no user-visible benefit today** — none of them
is on a light command path. The value is removing a false claim and making a
future light import safe by construction.

**Comment by AchuthReddy-16** [comment]
I'd like to take this if it is still free.

No claim comments on this thread that I can see. If that is right, I will follow the `trainer/raft.py` factory shape so the eleven callback modules import without pulling transformers/torch, keep the #308 `TrainerCallback` subclass property, and extend `tests/test_cli_startup_is_light.py` with a runtime per-module check.

I also asked on #280. Happy to do one or both, whichever you prefer.

**Comment by MakazhanAlpamys** [comment]
Yours — and sorry for the silence, that is fifteen hours you spent not knowing. Nobody else has claimed it; you read the thread correctly.

Your plan is right, and the `trainer/raft.py:187` shape is the correct reference: resolving the base **inside a factory function** is lazy, whereas `_TrainerCallbackBase = _get_trainer_callback_base()` and `class X(_try_import_callback_base())` both run at module import and cost ~4.4 s of transformers+torch.

Three things worth knowing before you start, because they are not visible from the issue text.

**The AST guard cannot see this defect, and that is the whole point.** A per-file `test_no_top_level_torch` proves a *syntactic* property — it cannot see a transitive import or an eagerly-called factory. The authority is `tests/test_cli_startup_is_light.py`, which asserts the *runtime* property (`torch` not in `sys.modules`) across all 88 command modules plus 11 real light-command invocations. Verify against that, not the AST guard.

**Measure before and after.** `soup --help` is the user-visible number. It regressed 5x once when a light command imported one of these modules (#320's own origin). Quote the timing both ways rather than asserting the import moved.

**#308 is the adjacent trap, not the same one.** Those callbacks must still subclass the real `TrainerCallback` — HF dispatches every event with no `hasattr` guard, so a duck-typed callback crashes on `on_epoch_begin`. Making the import lazy must not accidentally make the base class fake. If your change touches that, say so and I will look at it especially carefully.

Take it. If it turns out eleven modules need eleven different shapes, that is a finding worth reporting rather than forcing.

**Comment by MakazhanAlpamys** [comment]
Fixed by @AchuthReddy-16 in #455, merged as `b056aa8`.

Eleven callback modules pulled transformers at import because the lazy-LOOKING factory was called EAGERLY. Measured after: `soup_cli.utils.echo_trap` 4.30s -> 0.03s, `reward_hack_control` 3.71s -> 0.05s, `lisa` 3.46s -> 0.02s, with torch absent from `sys.modules`. Each class still resolves to a real `TrainerCallback` base, so #308's property is preserved.

**The part worth recording for whoever reads this issue next.** A first attempt deferred the classes but left the builders referencing bare globals, which PEP 562 `__getattr__` does not serve — so four of five builders raised `NameError` on a cold import, with ten `# noqa: F821` silencing the diagnostic. **CI reported one failure.** Eleven of the twelve vanish in a full-suite run, because an earlier test file fetches the attribute first and caches it into `globals()`.

So the guard that matters is `test_builder_invocation_no_nameerror_in_subprocess`, which spawns a fresh process per builder. Under a deliberate re-break, 37 in-process tests pass and only that one fails.

Try:

```
python -c "import soup_cli.utils.echo_trap, sys; print('torch' in sys.modules)"   # False
```
